src/main/java/server/managers/BookingManager.java

```diff
@@ -61,67 +60,275 @@ public static LocalTime getCloseTime() {
+        if (e != null){
+            return new Pair<Booking, Exception>(null, e);
+        }
+
+        //Booking procedure
+        Booking b = new Booking(day, bookingId, clientId, facilName, newStart, newEnd);
+        Facility f = (Facility) facilTable.get(facilName);
+        f.addBooking(b);
+        return new Pair<Booking, Exception>(b, null);
     }
 
-    public Pair<Booking, Exception> createBooking(int bookingId, int clientId, String facilName, LocalDateTime newStart, LocalDateTime newEnd, Hashtable facilTable){
+    public Pair<Booking, Exception> offsetBooking(String facilName, int bookingId, int offset, Hashtable facilTable){
+        //FIXME offset should be in minutes
+        Exception e;
 
-        facilName = facilName.toLowerCase();
+        Booking b = this.findFacilBookingById(facilName, bookingId, facilTable);
+        String day = b.getDay();
 
-        if (!facilTable.containsKey(facilName)){
-            Exception e = new NoSuchElementException("Facility does not exist"); //TODO review this exception
+        LocalTime newStart = b.getStart().plusMinutes(offset);
+        LocalTime newEnd = b.getEnd().plusMinutes(offset);
+
+        e = doBookingNotNullCheck(b);
+        if(e != null){
+            return new Pair<Booking, Exception>(null, e);
+        }
+        e = doBookingCheck(facilName,newStart,newEnd,facilTable);
+        if (e != null){
             return new Pair<Booking, Exception>(null, e);
         }
-        if (newStart.compareTo(newEnd) >= 0){
-            Exception e = new NoSuchElementException("Invalid start and end time"); //TODO review this exception
+        e = doAvailabilityCheckExceptCurrent(day, bookingId, facilName,newStart,newEnd,facilTable);
+        if (e != null){
             return new Pair<Booking, Exception>(null, e);
         }
+
+        b.setStart(newStart);
+        b.setEnd(newEnd);
+
+        return new Pair<Booking, Exception>(b, null);
+
+    }
+    public Pair<Booking, Exception> extendBooking(String facilName, int bookingId, int offset, Hashtable facilTable){
+        //FIXME offset should be in minutes
+        Exception e;
+        Booking b = findFacilBookingById(facilName, bookingId, facilTable);
+        LocalTime bStart = b.getStart();
+        LocalTime bEnd = b.getEnd();
+        LocalTime newEnd = bEnd.plusMinutes(offset);
+        String day = b.getDay();
+
+        e = doBookingNotNullCheck(b);
+        if(e != null){
+            return new Pair<Booking, Exception>(null, e);
+        }
+        e = doBookingCheck(facilName,bStart,newEnd,facilTable);
+        if (e != null){
+            return new Pair<Booking, Exception>(null, e);
+        }
+        e = doAvailabilityCheckExceptCurrent(day, bookingId, facilName,bStart,newEnd,facilTable);
+        if (e != null){
+            return new Pair<Booking, Exception>(null, e);
+        }
+
+        b.setEnd(newEnd);
+        return new Pair<Booking, Exception>(b, null);
+    }
+
+
```

Comment:
not an issue but just curious, how come Facility does not provide support for extend and offset booking? i.e. for create, query and delete, we use Facility's methods, but not for extend and offset

Reply:
wa good point, i think by my intended design (ie. Bookingmanager talks to Facility, Facility talks to Bookings/bookingslist) it should be what u mentioned.